satellite/contact/endpoint.go

```diff
@@ -37,46 +39,69 @@ func NewEndpoint(log *zap.Logger, service *Service) *Endpoint {
 func (endpoint *Endpoint) Checkin(ctx context.Context, req *pb.CheckinRequest) (_ *pb.CheckinResponse, err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	pingNodeSuccess, pingErrorMessage, err := pingBack(ctx, endpoint, req)
+	peerID, err := peerIDFromContext(ctx)
+	pingNodeSuccess, pingErrorMessage, err := pingBack(ctx, endpoint, req, peerID)
 	if err != nil {
 		return nil, Error.Wrap(err)
 	}
 
-	// Update the overlay cache with new uptime info from pinging back the node
-	peerID, err := peerIDFromContext(ctx)
+	// TODO(jg): We are making 2 requests to the database, one to update uptime and
+	// the other to update the capacity and operator info. We should combine these into
+	// one to reduce db connections. Consider adding batching and using a stored procedure.
 	_, err = endpoint.service.overlay.UpdateUptime(ctx, peerID, pingNodeSuccess)
 	if err != nil {
 		return nil, Error.Wrap(err)
 	}
 
-	// Should we update the req.NodeCapacity here as well?
+	nodeInfo := pb.InfoResponse{Operator: req.GetOperator(), Capacity: req.GetCapacity()}
+	_, err = endpoint.service.overlay.UpdateNodeInfo(ctx, peerID, &nodeInfo)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
 
 	return &pb.CheckinResponse{
 		PingNodeSuccess:  pingNodeSuccess,
 		PingErrorMessage: pingErrorMessage,
 	}, nil
 }
 
-func pingBack(ctx context.Context, endpoint *Endpoint, req *pb.CheckinRequest) (bool, string, error) {
+func pingBack(ctx context.Context, endpoint *Endpoint, req *pb.CheckinRequest, peerIDFromContext storj.NodeID) (bool, string, error) {
 	client, err := newClient(ctx,
 		endpoint.log,
 		endpoint.service.transport,
 		req.GetAddress(),
 	)
-	defer func() {
-		err = errs.Combine(err, client.close())
-	}()
 	if err != nil {
-		return false, "", err
+		// if this is a network error, then return the error otherwise just report internal error
+		_, ok := err.(net.Error)
+		if ok {
+			return false, "", Error.New("couldn't connect to client at addr %s due to network error: %v.", req.GetAddress().String(), err)
+		}
+		return false, "", Error.New("couldn't connect to client at addr: %s due to internal error.", req.GetAddress().String())
 	}
 
 	pingNodeSuccess := true
 	var pingErrorMessage string
 
-	_, err = client.pingNode(ctx, &pb.ContactPingRequest{})
+	p := &peer.Peer{}
+	_, err = client.pingNode(ctx, &pb.ContactPingRequest{}, grpc.Peer(p))
 	if err != nil {
 		pingNodeSuccess = false
-		pingErrorMessage = err.Error()
+		pingErrorMessage = "erroring while trying to pingNode due to internal error"
+		_, ok := err.(net.Error)
+		if ok {
+			pingErrorMessage = fmt.Sprintf("network erroring while trying to pingNode: %v\n", err)
+		}
+	}
+
+	// Confirm that the node ID from the initial checkin request
+	// matches that from this pingNode request
+	identityFromPeer, err := identity.PeerIdentityFromPeer(p)
+	if err != nil {
+		return false, "", Error.New("couldn't get identity from peer: %v", err)
+	}
+	if identityFromPeer.ID != peerIDFromContext {
```

Comment:
I don't think this works- this is comparing two slices, and not their contents. Maybe we need a `(*storj.NodeID).Equal()` method that does `bytes.Equal()`. (Or just do `bytes.Equal()` here)

Reply:
it looks like [storj.NodeID](https://github.com/storj/storj/blob/master/pkg/storj/node.go#L30) is an array and per [go docs](https://golang.org/ref/spec#Comparison_operators) "Array values are comparable if values of the array element type are comparable. Two array values are equal if their corresponding elements are equal."  So I think this comparison works, also a unit test I have just failed here so seems like its working.

Reply:
Apologies! I did not realize these were arrays, and I did not know that arrays were comparable. Carry on